core/src/state/world_state.rs

```diff
@@ -65,31 +65,73 @@ impl StateHandle {
             .write()
             .expect("Could not acquire world_state lock.")
     }
+}
 
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
-        }
-        let receiver = { self.write_state().get_listener(sequence) };
-        receiver.notified().await;
+/*
+We want a thing that waits until notified, and importantly waits
+for that notify on all of its replicas.
+The default notifier from tokio does not behave the way we need
+since, if multiple clones are made, notify_one only causes the first
+to return immediately when it invokes the notification and awaits.
+notify_waiters does not cause even the first one to do that.
+Therefore, we implement our own notifier that can be closed and
+will return immediately on all its dependent notifies.
+*/
+#[derive(Debug, Clone)]
+pub struct ClosableNotifier {
+    notifier: Sender<()>,
+}
+
+#[derive(Debug)]
+pub struct ClosableNotify(Receiver<()>);
+
+impl ClosableNotify {
+    pub async fn notified(&mut self) {
+        let _ = self.0.recv().await;
+    }
+}
+
+impl ClosableNotifier {
+    fn new() -> Self {
+        let (tx, _rx) = tokio::sync::broadcast::channel(128);
+        Self { notifier: tx }
+    }
+
+    fn notify(&self) -> ClosableNotify {
+        ClosableNotify(self.notifier.subscribe())
+    }
+
+    fn notify_waiters(&mut self) {
+        let _ = self.notifier.send(());
     }
 }
 
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
-    listeners: HashMap<u64, Arc<Notify>>,
+    listeners: RwLock<HashMap<u64, ClosableNotifier>>,
 }
 
 impl WorldState {
-    pub fn get_listener(&mut self, sequence: u64) -> Arc<Notify> {
-        match self.listeners.entry(sequence) {
-            Entry::Occupied(entry) => entry.get().clone(),
+    pub fn logical_time(&self) -> u64 {
+        self.logical_time
+    }
+
+    pub fn get_listener(&self, sequence: u64) -> ClosableNotify {
+        if self.logical_time >= sequence {
+            let mut notifier = ClosableNotifier::new();
+            let notify = notifier.notify();
```

Comment:
Is this clone necessary? `notifier` gets dropped anyway.

What if instead of `get_listener`, we combine this with the (former) async `wait_for_seq`, such that it is an async function. That way this whole branch could just be an immediate `return`, like it was in `wait_for_seq`.

Reply:
I'll fix that.

As for the second thing, making this async is iffy since I can't hold an RwLock across an await. I messed with this a little before rejecting it. Now I could use a tokio::RwLock, but then apply would have to become async, which is boo. (And DashMap, *especially* in this case, deadlocks like the US parliament.)

(I'm being a little mean, this is all in its documentation, but doesn't give any errors... probably PEBCAK tho)